src/ORM/DataQuery.php

```diff
@@ -1555,4 +1573,42 @@ private function withCorrectDatabase(callable $callback): mixed
         }
         return $callback();
     }
+
+    private function filterOrExcludeByQuery(DataQuery $subQuery, bool $isExclude, string $fieldToFilterBy, string $fieldInOtherList): static
+    {
+        if ($fieldInOtherList === '' || $fieldToFilterBy === '') {
+            throw new InvalidArgumentException('$fieldToFilterBy and $fieldInOtherList must not be empty strings');
+        }
+        $subSelect = $subQuery->getFinalisedQuery();
+        // Select only the relevant field, and ID. The ID is used below to validate the join.
+        $subSelect->setSelect([]);
+        $subSelect->selectField($subQuery->expressionForField($fieldInOtherList), $fieldInOtherList);
+        $subSelect->selectField($subQuery->expressionForField('ID'), 'ID');
+        // Remove order so the database can use whatever order is most efficient.
+        $subSelect->setOrderBy(null);
+
+        // Alias has to be unique so it doesn't conflict with another alias generated on the same query.
+        // A conflict is very unlikely but it's better to be 100% sure rather than having intermittent failures.
+        $from = $this->query->getFrom();
+        $subQueryAlias = null;
+        do {
+            // md5 is fast and very unlikely to have conflicts in this scenario.
+            $subQueryAlias = 'subQueryAlias' . RandomGenerator::singleton()->randomToken('md5');
+        } while (array_key_exists($subQueryAlias, $from));
+        // Join onto the subquery - this is more robust than adding it as a WHERE clause because it allows us to use an alias
+        // which protects us from errors if this method is called multiple times.
+        // We can't just do a straight "WHERE NOT IN (<subquery>)" because that doesn't account for null values, so any if we
+        // wanted to do that we'd need to use WHERE NOT EXISTS IN (<altered subquery>) where the altered subquery needs to check
+        // if the parent query $fieldToFilterBy is null-safe-equals to the subquery $fieldToFilterBy, and if both queries are on
+        // the same table that REQUIRES an alias to be used.
+        // tl;dr join works here while WHERE on its own would be complex to handle in an elegant way.
```

Comment:
```suggestion
        // tl;dr join on to the subquery works here while WHERE on its own would be complex to handle in an elegant way.
```
Move this to the top of the explanation

Reply:
Not sure what you mean? Do you mean put `tl;dr` at the top? If so.... how does the reader know where the `tl;dr` ends and the full explanation starts?

Have added the suggested "on to the subquery".

Reply:
tl;dr is more helpful at the top, because it's a summary and means people don't need to read the details. It's pretty normal to tl;dr's at the top of an article

> how does the reader know where the tl;dr ends and the full explanation starts?

`Explanation: ...`, or any other way that makes it obvious?

Reply:
Done.